### PR TITLE
Query Tool result export enhancements (JSON/XML, encoding, BOM, copy-with-headers)

### LINKED ISSUE #3205: Extend Download as CSV to include JSON and XML (RM #5045)
***Issue migrated from Redmine: https://redmine.postgresql.org/issues/5045***
*Originally created by **An Ne** at 2019-12-28 19:26:48 UTC.*

If a query returns a single row with a single column and is of type JSON or XML, Download to CSV should write the field as-is without header information using the appropriate file extension.

### LINKED ISSUE #4128: query tool csv/txt output file encoding choice (RM #6215)
***Issue migrated from Redmine: https://redmine.postgresql.org/issues/6215***
*Originally created by **alain dupeyron** at 2021-02-08 13:52:53 UTC.*

add a configuration parameter in pgadmin to select file encoding of query tool output file
default is utf-8, but I would like to choose my plateform encoding Windows-1252 by example

mail exchange on this subject below :

On Mon, Feb 8, 2021 at 5:20 AM Aditya Toshniwal <aditya.toshniwal@enterprisedb.com> wrote:
Hi Alain,
On Sat, Feb 6, 2021 at 3:20 PM <adupeyron@yahoo.fr> wrote:
It's why adding configuration parameter in pgadmin for file extension could solve the problem for most of people.And encoding in same time would complete the feature...You can raise the two separate feature requests here - https://redmine.postgresql.org/projects/pgadmin4/issues/new
Le 6 févr. 2021 00:17, Dave Caughey <caugheyd@gmail.com> a écrit :

On Fri, Feb 5, 2021 at 3:39 PM richard coleman <rcoleman.ascentgl@gmail.com> wrote:
Dave, 
I concur with Alain.
On Fri, Feb 5, 2021 at 3:22 PM Dave Caughey <caugheyd@gmail.com> wrote:
Always naming the file as ".csv", regardless of the delimited used, was a bug that was fixed.

In my case it was a bug that was created. The problem with always naming a file as ".csv" is that Excel then blindly assumes that the file actually does contain comma-delimited fields, and bypasses the Import From Text Wizard, so if you use anything other than commas (e.g., you want TAB separators), then Excel won't split your file into multiple columns correctly.   Therefore, before the bug was fixed, you had to manually rename all of pgAdmin's "export to file" outputs to ".txt".  (Exactly what you don't want to have to do, too!)

No, both Alain and myself now have to rename the .txt files to .csv files.  In my case it's because LibreOffice won't open it in Calc with a txt extension, but instead try to open it in Write.  If I leave it .txt I have to; manually start Calc, then select the file, then try to import it.  If it's named .csv a simple double-click opens the file in Calc and in the CSV import wizard that correctly handles files delimited by comma, tab, ~, |, ;, or anything else you like.
Yeah, that's definitely unfortunate, but I think accommodating Excel's lame behaviour far outweighs accommodating LibreOffice's lame behaviour, given the relative number of users of the two programs.
Clearly, this would be solvable by including a "default extension" (e.g., ".csv" or ".txt") that would probably make everyone happy, but it seems that simply going back to the old buggy behaviour would be worse for a lot more people than it would help.
 Note, if your file is ends with ".txt", then simply doing a "File" | "Open" in Excel will automatically/correctly invoke the Import from Text Wizard, and let you pick the delimiter you want (and it does a good job of auto-detecting).  Easy!

And, if you don't want to use commas as your delimiter (for any number of reasons), then you can change that in pgAdmin by going to "File" | "Preferences" | "CVS/TXT Output" and choose your preferred delimiter (e.g., Tab, "|", etc.), and thereafter that's what pgAdmin will always use when you do an "export to file", but will give it an ".txt" extension so Excel won't screw up opening the file because it's expecting commas as delimiters.

Currently the only options pgAdmin4 gives me are; semicolon, comma, vertical bar, or tab.  I can apparently just type whatever I want, but that's hardly obvious.  And comma is a very bad choice for European countries where sixteen hundred point fifty-two would be written as 1600,52 as opposed to 1600.52 on this side of the pond.  You might be suffering from just a touch of Americentric bias.   Cheers,
Dave

On Fri, Feb 5, 2021 at 1:25 PM Alain Dupeyron <adupeyron@yahoo.fr> wrote:

    Hi,

    I'm using frequently the download file function in the query tool. In previous version of PgAdmin4, I select field separator ";"  and file was always generated with extension .csv and can be opened directly by Excel in one click with correct columns !

    Now I have always .txt extension (since 4.28 or before, I'm in PgAdmin 4.29 for now) but the .txt file do not open directly in excel
    I Have to launch excel and open it but colums are not detected in my French excel, I have to convert data...(idem with TAB separator)...fastidious !
    Or I have to save the file from the browser, then navigate to file, then I change its extension to .csv, confirm and then the file is opened correctly...fastidious !

    When I choose field separator "," in PgAdmin, then file is generated with extension .csv but colums are not detected in French Excel (comma is the decimal number separator in France...)

    => How can I "force" or choose in Pgadmin to always generate file with extension .csv independently of selected field separator ?

    2nd point: Can we choose file encoding of generated file too ? it's in UTF-8 now (was plateform dependant in PgAdmin 3?)
    accentuated characters are not detected corretly by french excel (expecting by default my Windows-1252 plateform encoding), I have to import file in excel and to select the right utf8 encoding in a huge encoding list...

    Other approach : copy from query result then paste in excel and save the file myself...
    accent are ok but column header are missing, I need to select the option on each new query tab "copy with header" => there is no pgAdmin preferences for that point ? (was in PgAdmin 3)

    Yes we can blame Excel/Windows and csv format but it's the only download format proposed here  for spreadsheet tool and only theses 2-3 configurations are missing to parametrize it as expected by end user platform...

    Thanks

### LINKED ISSUE #4129: add query tool preferences for "copy with header" (RM #6216)
***Issue migrated from Redmine: https://redmine.postgresql.org/issues/6216***
*Originally created by **alain dupeyron** at 2021-02-08 13:57:26 UTC.*

I would be useful to add a query tool preferences to activate option "copy with header"

I need to select the option "copy with header" on EACH new query tab => there is no pgAdmin preferences for that point ? (was in PgAdmin 3)

### LINKED ISSUE #6695: Add a UTF BOM to the CSV produced by "Save Results to File" where appropriate
**Describe the solution you'd like**

When a database uses a UTF-8/16/32 encoding add the appropriate Byte Order Mark to the file produced by "Save Results To File"

**Describe alternatives you've considered**

* Tell every Excel loving data-scientist to always use the "Get Data From Text" wizard and select the right character encoding manually.
* Allow a per-client encoding configuration and use on-the-fly character set conversion to always export data in the configured encoding.

**Additional context**

Many windows applications assume the encoding is ANSI or ASCII unless a BOM is present.

### PR DESCRIPTION
**Pull request by dpage**
## Summary

A batch of long-standing Query Tool result export/copy enhancements, all in
the results download/copy path.

- **Save results as JSON or XML** (in addition to CSV), selectable from a
  drop-down on the *Save results to file* toolbar button. The download
  generator is now format-aware and streams JSON/XML as well as CSV. JSON/XML
  are always emitted as UTF-8; XML emits column names as escaped `name`
  attributes so column names that are not valid XML element names are handled
  safely.
- **Output file encoding** preference (Query Tool → CSV/TXT Output) controlling
  the character encoding used when saving results; defaults to utf-8, with a
  free-text option for encodings that are not listed.
- **Add byte order mark (BOM)?** preference that prepends a UTF BOM to saved
  CSV/TXT files for better interoperability with applications such as Microsoft
  Excel. (Applies to CSV/TXT output only.)
- **Copy with headers?** preference that seeds the default state of the results
  grid "Copy with headers" toggle (still toggleable per-copy).

## Testing

- New integration tests exercise the JSON, XML, BOM and non-UTF-encoding
  download paths through the real `/query_tool/download/` endpoint; the
  existing CSV scenarios continue to pass, confirming the generator refactor
  is non-regressive.
- `pycodestyle` and `eslint` clean.
- Preferences and Query Tool toolbar documentation updated, plus release-notes
  entries.

Closes #3205
Closes #4128
Closes #4129
Closes #6695

## Summary by CodeRabbit

* **New Features**
  * Download query results as CSV, JSON, or XML from the Query Tool.
  * Configure CSV/TXT output encoding and optional byte order marks (BOM).
  * Choose whether copied result data includes column headers by default.
  * Support improved filenames, separators, character handling, and data formatting across export types.

* **Documentation**
  * Updated Query Tool guidance for result downloads, output formats, encoding, BOM settings, and header copying.

### CODE REVIEW

**Comment by anthonydb** [comment]
@dpage You appear to be, as they say, on a roll.

**Comment by asheshv** [comment]
## Review

### Must-fix before merge

**1. `replace_nulls_with` corrupts typed nulls in JSON/XML**
`_generate_json` / `_generate_xml` apply `handle_null_values` the same as CSV. With the default `"NULL"`, JSON emits `{"col": "NULL"}` instead of `{"col": null}`, and XML's `null="true"` branch is preempted. Skip `replace_nulls_with` for `json`/`xml`.

**2. XML doesn't strip XML-1.0-invalid control chars**
`xml_escape(str(value))` lets `\x00`–`\x08`, `\x0B`, `\x0C`, `\x0E`–`\x1F` through. PG `text` can contain these, producing well-formed-looking but invalid XML that strict parsers reject. Filter them on the XML path.

**3. Silent data loss on restrictive encodings**
`chunk.encode(output_encoding, errors='replace')` turns un-representable chars (e.g. emoji → latin-1) into `?` with no indication. Either switch to `backslashreplace` for non-UTF, or document the replacement in the preference help text.

**Test gaps to close alongside:** JSON `SELECT NULL`, XML `SELECT chr(7)`, and latin-1 export of a non-Latin character.

### Nice to have

**1. `utf-8-sig` BOM inconsistency**
Self-emits a BOM even when `add_bom=False`. Either add `utf8sig` to the codec-self-emits set or note the quirk in the help text.

**2. Batch JSON output**
`_generate_json` yields one chunk per row; CSV batches by `records`. For million-row exports this is a lot of small WSGI yields — mirror the CSV `StringIO` batching.

**3. Remove unused `json_columns`**
Built in `gen()` but never read. Pre-existing dead code, but this refactor is the natural moment to drop it.

**4. Remember last export format**
The main download button is hardcoded to CSV; the dropdown choice isn't sticky. Persist last choice per-session so power users don't reopen the menu every time.

**5. Single source of truth for the default format**
`'csv'` is repeated as a default in `saveResultsToFile`, the `TRIGGER_SAVE_RESULTS` listener, and `downloadResult`. Extract a constant.

**Review by asheshv — CHANGES_REQUESTED** [review]
Security and encoding logic are sound (format allowlist, `codecs.lookup()` validation), but four correctness bugs need fixing before merge:

1. **XML output not well-formed on common PG data.** `xml.sax.saxutils.escape()` only escapes `<`, `>`, `&` — it passes through all XML 1.0-illegal control chars (U+0000–U+0008, U+000B, U+000C, U+000E–U+001F). A `text` / `varchar` column legally containing `chr(1)` etc. will produce a file that strict XML parsers reject outright. Need to sanitize / replace control chars (e.g. via regex → `�`) before `xml_escape`.
2. **`bytea` / `memoryview` columns produce garbage in JSON and XML.** psycopg3 returns bytea as `memoryview`; `_json_default` does `str(value)` which yields `<memory at 0x...>`. XML hits the same `str(value)` path. Add explicit `isinstance(value, (memoryview, bytes, bytearray))` handling — `.hex()` or base64.
3. **NaN / Infinity floats not valid JSON.** Python's `json.dumps` emits bare `NaN` / `Infinity` tokens by default; not RFC 7159, rejected by Jackson / Python `json.loads`. Either `allow_nan=False` + handle non-finite floats in `_json_default`, or stringify them before serialization.
4. **`Content-Disposition` filename not quoted.** `"attachment;filename={0}".format(filename)` breaks on filenames with spaces or special chars (RFC 6266 requires quoting). Fallback `download.csv` is hardcoded even for JSON / XML — should use `download.<extn>`.

Test gap (non-blocking but worth filing): no scenarios for NULL values in JSON/XML output, XML special chars in data, bytea columns in any new format, or NaN / Infinity in JSON.

File: web/pgadmin/utils/driver/psycopg3/connection.py [hunk 1]
```diff
@@ -54,6 +58,121 @@
...
+                    row_io.append(
+                        '<column name={0} null="true"/>'.format(
+                            _xml_attr(column)))
+                else:
+                    row_io.append('<column name={0}>{1}</column>'.format(
+                        _xml_attr(column), _xml_text(value)))
+            row_io.append('</row>')
+            yield ''.join(row_io)
+        results = cur.fetchmany(records)
+    yield '</data_output>'
```
[the remark this answers is not in the document]
Comment (dpage) [reply web/pgadmin/utils/driver/psycopg3/connection.py:173]: Fixed in 2b3c2fa9a: a genuine single-row, single-column result now yields the bare value (a JSON scalar, or an XML document with no <row>/<column> wrapper) instead of the collection wrapper, per #3205. Added JSON/XML regression tests, including the NULL single-value shape.

File: web/pgadmin/utils/driver/psycopg3/connection.py [hunk 2]
```diff
@@ -950,50 +1070,56 @@ def gen(conn_obj, trans_obj, quote='strings', quote_char="'",
...
-            res_io = StringIO()
-
-            if quote == 'strings':
-                quote = csv.QUOTE_NONNUMERIC
-            elif quote == 'all':
-                quote = csv.QUOTE_ALL
+            if data_format == 'json':
+                yield from _generate_json(cur, records, results)
+            elif data_format == 'xml':
+                yield from _generate_xml(cur, records, results, header)
```
[the remark this answers is not in the document]
Comment (dpage) [reply web/pgadmin/utils/driver/psycopg3/connection.py:1120]: Fixed in 2b3c2fa9a: the empty-result check now runs after the header is known and is format-aware, so a zero-row JSON/XML download returns an empty array / empty document rather than the plain-text "did not return any data" message under an application/json or application/xml content type. Added regression tests for both.

File: web/pgadmin/tools/sqleditor/tests/test_download_csv_query_tool.py [hunk 3]
```diff
@@ -240,3 +243,282 @@ def tearDown(self):
...
+                 awkward_data=True)
+        ),
+        (
+            'Download awkward data as XML stays well formed',
+            dict(data_format='xml', add_bom=False, encoding='utf-8',
+                 expected_content_type='application/xml',
+                 expected_extension='.xml', sql=AWKWARD_SQL,
+                 awkward_data=True)
+        ),
+    ]
```
[the remark this answers is not in the document]
Comment (dpage) [reply web/pgadmin/tools/sqleditor/tests/test_download_csv_query_tool.py:420]: Fixed in 2b3c2fa9a: added single-row, single-column scenarios (including a NULL case) for both JSON and XML, each asserting the direct-value contract (bare JSON scalar / XML with no row-column wrapper) rather than the collection shape.

File: web/pgadmin/tools/sqleditor/tests/test_download_csv_query_tool.py [hunk 4]
```diff
@@ -240,3 +243,282 @@ def tearDown(self):
...
+            'Download CSV with a UTF BOM',
+            dict(data_format='csv', add_bom=True, encoding='utf-8',
+                 expected_content_type='text/csv',
+                 expected_extension='.csv')
+        ),
+        (
+            'Download CSV with a non-UTF output encoding',
+            dict(data_format='csv', add_bom=True, encoding='latin-1',
+                 expected_content_type='text/csv',
+                 expected_extension='.csv')
```
[the remark this answers is not in the document]
Comment (dpage) [reply web/pgadmin/tools/sqleditor/tests/test_download_csv_query_tool.py]: Fixed in 2b3c2fa9a: the Latin-1 output-encoding scenario now uses '€' (a character Latin-1 cannot represent) instead of ASCII-only data, and asserts the exact defined behaviour: the exporter's existing errors='replace' contract turns it into '?' rather than dropping or corrupting it.

**Comment by dpage** [comment]
Thanks for the thorough review, @asheshv. Checked all four against the current tip of the branch:

1. **XML control chars** — already fixed: `connection.py` now runs values through an `_XML_ILLEGAL_CHARS` regex (everything outside the XML 1.0-legal ranges) and substitutes U+FFFD before `xml_escape`, both for element text and attribute values.
2. **bytea/memoryview in JSON/XML** — already fixed: a shared `_to_text()` helper explicitly checks `isinstance(value, (memoryview, bytes, bytearray))` and renders it as `\x`-prefixed hex, used by both `_json_safe()` (JSON) and `_xml_text()` (XML).
3. **NaN/Infinity in JSON** — already fixed: `_json_safe()` walks dicts/lists/tuples and rewrites any non-finite float to the `'NaN'`/`'Infinity'`/`'-Infinity'` string PostgreSQL itself uses, and `json.dumps(..., allow_nan=False)` is kept as a backstop so nothing bare can slip through.
4. **Content-Disposition quoting/extension** — already fixed: the filename is always RFC 6266-quoted (backslash/quote escaped), with an RFC 5987 `filename*=UTF-8''...` fallback for non-Latin-1 names, and the ASCII stand-in is `download.<extn>` matching the actual export format rather than a hardcoded `.csv`.

All four landed in the "Fix double BOM..." and "Make the JSON and XML output parseable..." commits already on the branch, so no new code changes were needed this round.

Test gap: added `TestDownloadResultFormats` scenarios in `test_download_csv_query_tool.py` covering exactly what you flagged — a control character (`chr(1)`) plus a `bytea` column plus `NaN`/`Infinity`/`NULL` in one row, asserted via `json.loads(..., parse_constant=...)` (rejecting bare NaN/Infinity) and `ElementTree.fromstring()` (rejecting the control char / malformed XML) for both JSON and XML. Also added coverage for single-value results, empty result sets, and the two filename-quoting edge cases (spaces, non-Latin-1).

Rebased onto current upstream/master and re-ran the full module locally (22/22 passing). Ready for another look whenever you get a chance.

File: web/pgadmin/tools/sqleditor/tests/test_download_csv_query_tool.py [hunk 5]
```diff
@@ -240,3 +243,390 @@ def tearDown(self):
...
+            # two (which happened when a BOM was hand-prepended for codecs
+            # that already self-emit one, e.g. utf-16/utf-32).
+            bom = {
+                'utf8': codecs.BOM_UTF8,
+                'utf16': codecs.BOM_UTF16,
+                'utf32': codecs.BOM_UTF32,
+            }[normalized]
+            self.assertTrue(raw.startswith(bom))
+            # No second, redundant BOM immediately after the first.
+            self.assertFalse(raw[len(bom):].startswith(bom))
```
[the remark this answers is not in the document]
Comment (dpage) [reply web/pgadmin/tools/sqleditor/tests/test_download_csv_query_tool.py:598]: Good catch, and it was hiding a real bug rather than just a weak test. The exporter fetches ten rows at a time (`execute_on_server_as_csv(records=10)`) and encoded each chunk with its own `chunk.encode()` call, and the `utf-16`/`utf-32` codecs emit their BOM on every such call, so any export spanning more than one chunk carried a BOM at the head of every chunk. The single row fixture could never reach a second chunk, so nothing noticed.

Fixed in 15ec50e90: encoding now goes through `codecs.getincrementalencoder()`, which emits the BOM once and then keeps going, with a final flush for anything the codec is still holding. There is a new 25 row utf-16 scenario alongside the existing one, and the assertion now counts BOMs across the whole payload rather than only checking that a second does not immediately follow the first. The fixture data is ASCII, so the BOM byte sequence cannot occur in the encoded rows and the count is exact.

I checked the new test fails without the encoder change and passes with it.

File: web/pgadmin/tools/sqleditor/tests/test_download_csv_query_tool.py [hunk 6]
```diff
@@ -240,3 +243,390 @@ def tearDown(self):
...
+                # errors='replace' must turn the character Latin-1 cannot
+                # encode into the codec's standard replacement rather than
+                # silently dropping it or corrupting the row.
+                self.assertNotIn(self.non_latin1_char, body)
+                self.assertIn('?', body)
+
+        url = '/sqleditor/close/{0}'.format(self.trans_id)
+        response = self.tester.delete(url)
+        self.assertEqual(response.status_code, 200)
+        database_utils.disconnect_database(self, self._sid, self._did)
```
[the remark this answers is not in the document]
Comment (dpage) [reply web/pgadmin/tools/sqleditor/tests/test_download_csv_query_tool.py:643]: Agreed, and moved in 15ec50e90. Closing the transaction and disconnecting now happen in `tearDown` ahead of `test_utils.drop_database`, so a failed assertion no longer skips them and leaves the connection open against a database that is about to be dropped. The close is no longer asserted on, since a teardown is the wrong place to turn a cleanup hiccup into a second, confusing failure on top of the real one.

**Comment by dpage** [comment]
@asheshv, I've been back through your earlier review comment as well, since my last reply only covered the four points in the formal review. Rebased onto current master, with the remainder in 29982a310:

**Must-fix**
1. `replace_nulls_with` in JSON/XML: already fixed in 723af549e; NULLs come out as JSON `null` and XML `null="true"`, and the preference applies to CSV only.
2. XML control characters: already fixed in the same commit (replaced with U+FFFD), with a `chr(1)` test.
3. Silent loss on restrictive encodings: I've kept `errors='replace'` and documented it, in both the preference help text and `preferences.rst`. `backslashreplace` would put `€` into a CSV cell, which looks like data rather than a substitution, and the latin-1 `€` test pins the `?` behaviour.

**Nice to have**
1. `utf-8-sig`: this was worse than an inconsistency, as with the BOM switch on it produced two BOMs. It's now treated like `utf-16`/`utf-32` (never hand-prepended), with a test, and the help text says those three always carry a BOM.
2. JSON (and XML) now yield one chunk per fetched batch, as CSV does; there are 25-row scenarios for both, checking that every row comes through across the batch boundaries.
3. `json_columns` is gone, along with the `ALL_JSON_TYPES` import that only it used.
5. The `'csv'` default is now `DEFAULT_RESULT_DOWNLOAD_FORMAT` in `QueryToolConstants.js`.

I've left out 4 (remembering the last export format). It changes what the main button and F8 do based on a choice made earlier in a menu, and it would need a visible cue as to which format the button will now produce. That seems like a follow-up worth doing properly rather than something to bolt on here; I'm happy to open an issue for it.

File: web/pgadmin/tools/sqleditor/__init__.py [hunk 7]
```diff
@@ -2179,35 +2180,120 @@ def start_query_download_tool(trans_id):
...
+            '-', '').replace('_', '')
+        is_utf = normalized_encoding.startswith('utf')
+        # The 'utf-16' and 'utf-32' codecs (without an explicit endianness
+        # suffix) and 'utf-8-sig' emit their own BOM, so we must not
+        # hand-prepend one too; doing so would produce two BOMs and corrupt
+        # the output. The explicit-endian forms (utf-16-le/-be,
+        # utf-32-le/-be) and utf-8 do not self-emit a BOM, so for those we
+        # keep writing it ourselves.
+        codec_self_emits_bom = normalized_encoding in (
+            'utf16', 'utf32', 'utf8sig')
```
[the remark this answers is not in the document]
Comment (dpage) [reply web/pgadmin/tools/sqleditor/__init__.py]: Fixed in 458dbc353: the codec's canonical name now drives the BOM decisions and the charset header, with scenarios for 'u8' and 'utf-16 '.

File: web/pgadmin/utils/driver/psycopg3/connection.py [hunk 8]
```diff
@@ -54,6 +58,143 @@
...
+    """
+    if data_format == 'json':
+        return json.dumps(
+            _json_safe(value), default=_json_default, allow_nan=False)
+
+    if value is None:
+        return ('<?xml version="1.0" encoding="UTF-8"?>\n'
+                '<data_output null="true"/>')
+    return ('<?xml version="1.0" encoding="UTF-8"?>\n'
+            '<data_output>{0}</data_output>'.format(_xml_text(value)))
```
[the remark this answers is not in the document]
Comment (dpage) [reply web/pgadmin/utils/driver/psycopg3/connection.py]: Fixed in 458dbc353: a lone json/jsonb value exported as JSON is written as is, and so is a lone xml value exported as XML, which #3205 asks for too. Tests cover all three.
